### PR TITLE
ENH: keepdims=True for xarray reductions

### LINKED ISSUE #2170: keepdims=True for xarray reductions
For operations where arrays are aggregated but then combined, the `keepdims=True` option for NumPy aggregations is convenient.

We should consider supporting this in xarray as well. Aggregating a DataArray/Dataset with `keepdims=True` (or maybe `keep_dims=True`) would remove all original coordinates along aggregated dimensions and return a result with a dimension of size 1 without any coordinates, e.g.,
```
>>> array = xr.DataArray([1, 2, 3], dims='x', coords={'x': ['a', 'b', 'c']})
>>> array.mean(keepdims=True)
<xarray.DataArray (x: 1)>
array([2.])
Dimensions without coordinates: x
```

In case, `array.mean(keepdims=True()` is equivalent to `array.mean().expand_dims('x')` but in general this equivalent does not hold, because the location of the original dimension is lost.

Implementation-wise, we have two options:
1. Pass on `keepdims=True` to NumPy functions like `numpy.mean()`, or
2. Implement `keepdims=True` ourselves, in `Variable.reduce()`.

I think I like option 2 a little better, because it places fewer requirements on aggregation functions. For example, functions like `bottleneck.nanmean()` don't accept a `keepdims` argument.

### PR DESCRIPTION
**Pull request by ScottWales**
Add new option `keepdims` to xarray reduce operations, following the behaviour of Numpy.

`keepdims` may be passed to reductions on either Datasets or DataArrays, and will result in the reduced dimensions being still present in the output with size 1.

Coordinates that depend on the reduced dimensions will be removed from the Dataset/DataArray

The name `keepdims` is used here to be consistent with Numpy, `keep_dims` was an alternative name proposed in #2170.

The functionality has only been added to `Variable.reduce()`, `Dataarray.reduce()` and `DataSet.reduce()` to start off with, it is not implemented for `GroupBy`, `Rolling` or `Resample`.

 - [x] Closes #2170
 - [x] Tests added
 - [x] Fully documented, including `whats-new.rst` for all changes and `api.rst` for new API

### CODE REVIEW

File: xarray/core/variable.py [hunk 1]
```diff
@@ -1334,7 +1334,7 @@ def where(self, cond, other=dtypes.NA):
         return ops.where_method(self, cond, other)
 
     def reduce(self, func, dim=None, axis=None,
-               keep_attrs=None, allow_lazy=False, **kwargs):
```
Comment (max-sixty) [inline xarray/core/variable.py]: Is there a reason this isn't default `False`?
(I'm not sure why `keep_attrs` is default `None`, but looks like it's calling into `_get_keep_attrs` at least)

Comment (ScottWales) [reply xarray/core/variable.py]: Yes it was to follow the behaviour of `keep_attrs`, happy to change the default to be `False`

File: xarray/core/dataarray.py [hunk 2]
```diff
@@ -1657,6 +1664,10 @@ def reduce(self, func, dim=None, axis=None, keep_attrs=None, **kwargs):
             If True, the variable's attributes (`attrs`) will be copied from
             the original object to the new one.  If False (default), the new
             object will be returned without attributes.
+        keepdims : bool, optional
```
Comment (max-sixty) [inline xarray/core/dataarray.py]:
```suggestion
        keepdims : bool, default False
```

File: xarray/core/dataarray.py [hunk 3]
```diff
@@ -1637,7 +1643,8 @@ def combine_first(self, other):
         """
         return ops.fillna(self, other, join="outer")
 
-    def reduce(self, func, dim=None, axis=None, keep_attrs=None, **kwargs):
+    def reduce(self, func, dim=None, axis=None, keep_attrs=None,
```
Comment (max-sixty) [inline xarray/core/dataarray.py]:
```suggestion
    def reduce(self, func, dim=None, axis=None, keep_attrs=False,
```

File: xarray/core/dataset.py [hunk 4]
```diff
@@ -3149,6 +3149,10 @@ def reduce(self, func, dim=None, keep_attrs=None, numeric_only=False,
             If True, the dataset's attributes (`attrs`) will be copied from
             the original object to the new one.  If False (default), the new
             object will be returned without attributes.
+        keepdims : bool, optional
```
Comment (max-sixty) [inline xarray/core/dataset.py]:
```suggestion
        keepdims : bool, default False
```

File: xarray/core/variable.py [hunk 5]
```diff
@@ -1354,6 +1354,9 @@ def reduce(self, func, dim=None, axis=None,
             If True, the variable's attributes (`attrs`) will be copied from
             the original object to the new one.  If False (default), the new
             object will be returned without attributes.
+        keepdims : bool, optional
```
Comment (max-sixty) [inline xarray/core/variable.py]:
```suggestion
        keepdims : bool, default False
```

File: xarray/core/dataarray.py (hunk 3 above)
Comment (ScottWales) [reply xarray/core/dataarray.py]: This seems outside the scope of the current change

**Comment by ScottWales** [comment]
Thanks for the review @max-sixty, happy to help out

File: xarray/core/variable.py [hunk 6]
```diff
@@ -1388,6 +1391,12 @@ def reduce(self, func, dim=None, axis=None,
             keep_attrs = _get_keep_attrs(default=False)
         attrs = self._attrs if keep_attrs else None
 
+        if keepdims:
+            for i, d in enumerate(self.dims):
+                if d not in dims:
+                    dims.insert(i, d)
+                    data = np.expand_dims(data, axis=i)
```
Comment (shoyer) [inline xarray/core/variable.py]: This works on NumPy arrays, but won't work on dask arrays (or other "duck array" types that we want to support in the future).

Two alternatives that would work here:
1. define `expand_dims` in `duck_array_ops.py` using `_dask_or_eager_func`.
2. expand dimensions by indexing with a tuple built-up out of `np.newaxis`/`slice(None)` instead.

File: xarray/core/dataarray.py (hunk 3 above)
Comment (max-sixty) [reply xarray/core/dataarray.py]: Ah, wrong one. Thanks

File: xarray/core/variable.py (hunk 6 above)
Comment (ScottWales) [reply xarray/core/variable.py]: Thanks @shoyer, I've added option 2 and a test for dask variables

File: xarray/core/variable.py [hunk 7]
```diff
@@ -1388,6 +1391,12 @@ def reduce(self, func, dim=None, axis=None,
             keep_attrs = _get_keep_attrs(default=False)
         attrs = self._attrs if keep_attrs else None
 
+        if keepdims:
```
[the remark this answers is not in the document]
Comment (ScottWales) [reply xarray/core/variable.py]: This has been moved as suggested

File: xarray/core/variable.py [hunk 8]
```diff
@@ -1381,8 +1384,19 @@ def reduce(self, func, dim=None, axis=None,
         else:
             removed_axes = (range(self.ndim) if axis is None
                             else np.atleast_1d(axis) % self.ndim)
-            dims = [adim for n, adim in enumerate(self.dims)
-                    if n not in removed_axes]
+            if keepdims:
+                # Insert np.newaxis for removed dims
+                slices = tuple(np.newaxis if i in removed_axes else
+                               slice(None, None) for i in range(self.ndim))
+                if getattr(data, 'shape', None) is None:
```
Comment (shoyer) [inline xarray/core/variable.py]: I think this would be equivalent to just `hasattr(data, 'shape')`?

File: xarray/core/variable.py [hunk 9]
```diff
@@ -1381,8 +1384,19 @@ def reduce(self, func, dim=None, axis=None,
         else:
             removed_axes = (range(self.ndim) if axis is None
                             else np.atleast_1d(axis) % self.ndim)
-            dims = [adim for n, adim in enumerate(self.dims)
-                    if n not in removed_axes]
+            if keepdims:
+                # Insert np.newaxis for removed dims
+                slices = tuple(np.newaxis if i in removed_axes else
+                               slice(None, None) for i in range(self.ndim))
```
Comment (shoyer) [inline xarray/core/variable.py]: `slice(None)` is equivalent to `slice(None, None)`

**Comment by shoyer** [comment]
Thanks @ScottWales!

**Pull request merged**